File: codes/rust/76.最小覆盖子串.rs
/*
 * @lc app=leetcode.cn id=76 lang=rust
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let mut need_map: HashMap<char, i32> = HashMap::new();
        let mut window_map: HashMap<char, i32> = HashMap::new();
        let mut min_window_size = s.len();
        let mut min_left: i32 = -1;
        let mut min_right: i32 = -1;
        let mut valid_num = 0;
        let (mut left, mut right) = (0, 0);
        // 初始化需要覆盖字符的信息
        for c in t.chars() {
            *need_map.entry(c).or_insert(0) += 1;
        }
        let s_chars = s.chars().collect::<Vec<char>>();
        // 移动右游标扩大窗口
        while right < s_chars.len() {
            let ch = s_chars[right];
            right += 1;

            *window_map.entry(ch).or_insert(0) += 1;
            if need_map.contains_key(&ch) && window_map[&ch] == need_map[&ch] {
                valid_num += 1;
            }

            // 还未覆盖，继续扩大右窗口
            if valid_num < need_map.len() {
                continue;
            }

            // 更新最小的覆盖子串长度大小和左、右游标
            if right - left <= min_window_size {
                min_window_size = right - left;
                min_left = left as i32;
                min_right = right as i32;
            }

            // 已经覆盖了，开始移动左游标缩短窗口
            while left < right && valid_num == need_map.len() {
                let ch = s_chars[left];
                left += 1;
                if need_map.contains_key(&ch) && window_map[&ch] == need_map[&ch] {
                    valid_num -= 1;
                }
                if window_map.contains_key(&ch) {
                    *window_map.get_mut(&ch).unwrap() -= 1;
                }

                // 继续判断是否完全覆盖
                if valid_num == need_map.len() && right - left <= min_window_size {
                    min_window_size = right - left;
                    min_left = left as i32;
                    min_right = right as i32;
                }
            }
        }
        if min_left != -1 && min_right != -1 {
            let result = s[min_left as usize..min_right as usize].to_string();
            return result;
        }
        "".to_string()
    }
}
// @lc code=end

File: codes/golang/76.最小覆盖子串.go
/*
 * @lc app=leetcode.cn id=76 lang=golang
 *
 * [76] 最小覆盖子串
 */

 import "math"

// @lc code=start
func minWindow(s string, t string) string {
	need := make(map[byte]int)
	window := make(map[byte]int)

	// 先缓存字符串t中的字符和出现次数
	for i:=0; i < len(t); i++ {
		need[t[i]]++
	}
	// fmt.Printf("need: %v \n",need)
	left,right,valid := 0,0,0
	start,size := 0,math.MaxInt32

	for right < len(s) {

		c := s[right]

		right++ // 右移窗口
		//右移过程中对t中的数据出现次数累加
		if _,ok := need[c]; ok {
			window[c]++
			// 如果need中的出现次数和窗口中的出现次数相同的话就表示已经覆盖了
			if window[c] == need[c] {
				valid++
			}
		}

		// 判断是否已经覆盖完字串了
		for valid == len(need) {
			// 更新窗口起始位置和大小
			if right - left < size {
				start = left
				size = right - left
			}
			// 取出要移除掉的字符
			d := s[left]
			left++
			// 如果移除掉的字符存在
			if _,ok :=need[d]; ok {
				// 开始不含字串了，把valid值减1
				if window[d] == need[d] {
					valid--					
				}
				window[d]--
			}
		}
	}
	// 返回最小覆盖子串
	if size == math.MaxInt32 {
		return ""
	}
	return s[start:(start+size)]
}
// @lc code=end

